src/core/thread/mle.cpp

```diff
@@ -4741,5 +4764,49 @@ void Mle::DelayedResponseMetadata::RemoveFrom(Message &aMessage) const
     SuccessOrAssert(aMessage.SetLength(aMessage.GetLength() - sizeof(*this)));
 }
 
+Error Mle::DetachGracefully(otDetachGracefullyCallback aCallback, void *aContext)
+{
+    Error error = kErrorNone;
+
+    VerifyOrExit(aCallback != nullptr, error = kErrorInvalidArgs);
```

Comment:
The callback can be nullptr when the caller doesn't care when this is done.

Reply:
We are using the value of callback to determine if detaching is already in progress. Is it better to use another variable to save the state?

Reply:
If the callback is required, we should highlight it in the API description. However, from a user perspective, I guess sometimes the caller doesn't want to handle the done event. So, maybe we should allow nullptr for the callback. To determine the detaching state, can we add a new state detaching?

/cc/ @jwhui @abtink

Reply:
Agree with @bukepo . It makes sense for users to not provide callbacks. The device would enter `disabled` state anyhow after a short period of time.

Reply:
Adding "detaching" to `otDeviceRole` sounds too much change. It'd require changes to almost every other component.

I think we can add it to some MLE-internal state, like `AttachState`.

Reply:
We may simply use `mDetachGracefullyTimer.IsRunning()` to indicate if the device is detaching gracefully?

Reply:
`mDetachGracefullyTimer.IsRunning()` sounds good. Done.